Entities/Attachable.cpp

```diff
@@ -1,679 +1,450 @@
+			for (AEmitter *wound : m_Wounds) {
+				totalDamage += wound->CollectDamage();
+			}
+			for (Attachable *attachable : m_Attachables) {
+				totalDamage += attachable->CollectDamage();
+			}
+			return totalDamage * m_DamageMultiplier;
 		}
-		else if (IsCollidingWithTerrainWhileAttached() && !enable)
-		{
-			m_pParent->GetAtomGroup()->RemoveAtoms(GetAtomSubgroupID());
-			SetIsCollidingWithTerrainWhileAttached(false);
+		return 0;
+	}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	void Attachable::SetCollidesWithTerrainWhileAttached(bool collidesWithTerrainWhileAttached) {
+		if (m_CollidesWithTerrainWhileAttached != collidesWithTerrainWhileAttached) {
+			AddOrRemoveAtomsFromRootParentAtomGroup(collidesWithTerrainWhileAttached);
+			m_CollidesWithTerrainWhileAttached = collidesWithTerrainWhileAttached;
 		}
 	}
-	else if (IsAttached() && !CanCollideWithTerrainWhenAttached())
-	{
-		if (enable || !enable)
-		{
-			g_ConsoleMan.PrintString("ERROR: Tried to toggle collisions for attachable that was not set to collide when initially created!");
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	bool Attachable::ParticlePenetration(HitData &hd) {
+		bool penetrated = MOSRotating::ParticlePenetration(hd);
+
+		if (hd.Body[HITOR]->DamageOnCollision() != 0) {
+			AddDamage(hd.Body[HITOR]->DamageOnCollision());
 		}
+
+		if (penetrated && m_Parent) {
+			if (hd.Body[HITOR]->DamageOnPenetration() != 0) {
+				AddDamage(hd.Body[HITOR]->DamageOnPenetration());
+			}
+
+			// If the parent is an actor, generate an alarm point for them, moving it slightly away from the body (in the direction they got hit from) to get a good reaction.
+			Actor *parentAsActor = dynamic_cast<Actor *>(GetRootParent());
+			if (parentAsActor) {
+				Vector extruded(hd.HitVel[HITOR]);
+				extruded.SetMagnitude(parentAsActor->GetHeight());
+				extruded = m_Pos - extruded;
+				g_SceneMan.WrapPosition(extruded);
+				parentAsActor->AlarmPoint(extruded);
+			}
+		}
+
+		return penetrated;
 	}
-}
 
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 
-//////////////////////////////////////////////////////////////////////////////////////////
-// Method:          TransferJointForces
-//////////////////////////////////////////////////////////////////////////////////////////
-// Description:     Bundles up all the accumulated forces of this Attachable and calcs
-//                  how they transfer to the joint, and therefore to the parent.
+	void Attachable::GibThis(const Vector &impactImpulse, MovableObject *movableObjectToIgnore) {
+		MOSRotating::GibThis(impactImpulse, movableObjectToIgnore);
+		if (m_Parent) {
+			m_Parent->RemoveAttachable(this, true, true);
+		} else {
+			SetParent(nullptr);
```

Comment:
Good point! This was needed before when things were done more crappily, but now it's entirely pointless since SetParent would just early return anyway cause m_Parent == newParent.